### PR TITLE
Modify relevant NVIDIA control panel settings on Windows

### PR DESCRIPTION
**Pull request by ns6089**
## Description
This automates NVIDIA control panel settings modification, namely:
- sunshine.exe power profile, this can easily cut down 15ms of desktop latency, because default settings keep gpu clocks (particularly memory) at bare minimum
- global OpenGL/Vulkan DXGI swapchain, required for full frame rate capture with these apis, the setting is reverted to default value on sunshine.exe exit

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by ns6089** [comment]
@ReenigneArcher So, regarding nvapi submodule. Could you list exact steps that must be taken there (and who should perform them).

File: tools/nvprefs/nvprefs.cpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+  NvAPI_Status
+  restore_global_profile(NvDRSSessionHandle session_handle, bool &session_modified) {
+    NvAPI_Status status;
+
+    NvDRSProfileHandle profile_handle = 0;
+    status = NvAPI_DRS_GetBaseProfile(session_handle, &profile_handle);
+    if (status != NVAPI_OK) return status;
+
+    status = NvAPI_DRS_RestoreProfileDefaultSetting(session_handle, profile_handle, OGL_CPL_PREFER_DXPRESENT_ID);
```
Comment (cgutman) [inline tools/nvprefs/nvprefs.cpp:138]: We probably want to remember that we changed this and only restore the previous setting if we changed it. Otherwise we might revert this setting to default after a user changed it manually.

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:138]: Will work until the first crash or otherwise improper system shutdown. Pretty much all systems will normalize to "enabled" sooner or later.

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:138]: I guess we can save it to file?

Comment (cgutman) [reply tools/nvprefs/nvprefs.cpp:138]: Yeah, a file or a registry key seems reasonable.

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:138]: Still, one little mistake and we pollute the settings with false "enabled", many ways it can go wrong. Resetting to default is simple, and should be reasonably safe to merge this close to release. I believe anything more complex needs more time in nightly.

We can reset to default this release, and leave the complex logic for the next. Or we can simply move the whole thing to the next. Thoughts?

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:138]: *Still*, we can similarly pollute the system during uninstallation, if it doesn't terminate sunshine.exe gracefully. Let's move it to the next release after all, too many potential problems.

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:138]: The things to keep track of (so far)
 - use registry, to persist across reinstallations, especially portable
 - call restore from the uninstaller, in case sunshine didn't exit gracefully
 - save "using default" as separate value, in case nvidia decides to change it

File: tools/nvprefs/nvprefs.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+    return NVAPI_OK;
+  }
+
+  NvAPI_Status
+  restore_global_profile(NvDRSSessionHandle session_handle, bool &session_modified) {
+    NvAPI_Status status;
+
+    NvDRSProfileHandle profile_handle = 0;
+    status = NvAPI_DRS_GetBaseProfile(session_handle, &profile_handle);
```
Comment (Lucide) [inline tools/nvprefs/nvprefs.cpp:135]: You might want to use [`NvAPI_DRS_GetCurrentGlobalProfile()`](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#ga68dbe5987a8f530a0041ba94019b0ec0) instead. See [here](https://github.com/Lucide/vsync-tool#nvdrs_settingsettinglocation-values).

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:135]: From [NVIDIA Driver Settings Programming Guide](https://developer.download.nvidia.com/tools/docs/PG-5116-001_v02_public.pdf):
> A profile with one or more applications is called an Application Profile. Applications
that have a profile have their settings applied automatically when the application starts.
> 
> The Base Profile is a profile that always exists and is not associated with a specific
application. The settings in the Base Profile are applied to all processes on the system
automatically.
> 
> A profile with no applications is called a Global Profile. The settings from a Global
Profile are applied to all processes on the system, but only if that profile is selected to be
the Current Global Profile in the system.

...

> When an application starts, there might be three profiles contesting to provide settings to
the NVIDIA driver: the Application Profile, the Current Global Profile and the Base
Profile.
> 
> The settings from the Application Profile have the highest priority and override any
settings on the other two profiles. Next in priority order comes the Current Global
Profile, and finally the Base Profile.

No idea if anyone actually uses global profiles, but if somebody changes the profile we will lose the setting, even if the profile doesn't change that setting to begin with. There might be some sense in changing both base and global profiles at the same time, but it's going to complicate the undo logic, I will think about it.

File: tools/nvprefs/nvprefs.cpp (hunk 1 above)
Comment (Lucide) [reply tools/nvprefs/nvprefs.cpp:138]: Some potential useful things to keep in mind:

* nvapi drs api offers [`NvAPI_DRS_SaveSettingsToFile()`](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#gaf2f69fee6df6510c8cb5a5468b0e0b05), this call makes a full copy of the settings db to a file. The backup file can be loaded and navigated the same way as the actual settings.
* I agree that restoring previous settings is important. Nvidia driver settings use inheritance so this is not so trivial. 
  * In our case, we can probably assume that sunshine will have no "bundled" app profile (an app profile bundled with the driver). This allows us to only use `.settingLocation != NVDRS_CURRENT_PROFILE_LOCATION` to determine if a setting has to be set to a previously user value or reset. (In a general case, we would also need to check `.isCurrentPredefined`).
  * _Getting_ the current setting value can fail if the setting is inherited. Just a reminder.
*  That said, restoring a custom setting for the sunshine profile is probably unnecessary. The global setting must be restored. With the assumption above, the restore process uses the same set-or-reset logic.
* As you wrote above, guaranteeing restoration is complicated. Avoiding global profile tampering altogether might simply be better? The user can use the tool as a do-undo command.

File: tools/nvprefs/nvprefs.cpp (hunk 2 above)
Comment (Lucide) [reply tools/nvprefs/nvprefs.cpp:135]: `NVDRS_GLOBAL_PROFILE_LOCATION` points to the active global profile. In most cases, yes, it will be the base profile. If the user uses something like [NVIDIA Profile Inspector](https://github.com/Orbmu2k/nvidiaProfileInspector), that might not be the case. You do not need to keep track of both, only the global profile will be sufficient.

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp:135]: Global profile doesn't replace base profile, it extends it. It can't "point" to base profile. Similarly to application profile. Basically both global and application profiles work on opt-in basis and only override specific settings.

And regarding NVIDIA Profile Inspector, I don't see the ability to *change* global profile. It can change base profile, create named profile, and add applications to it.

Comment (Lucide) [reply tools/nvprefs/nvprefs.cpp:135]: I think I remember that, in a "standard" situation, a global profile might not exist. In this case, `NvAPI_DRS_GetCurrentGlobalProfile()` returns the same as `NvAPI_DRS_GetBaseProfile()`, and `.settingLocation` is always `NVDRS_GLOBAL_PROFILE_LOCATION`. A different current global profile can be set with [`NvAPI_DRS_SetCurrentGlobalProfile()`](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#ga35f52bf2ffa9d2a8f758912c41f4e936). You're right, in the profile inspector, you can't set the current global profile, but if you do programmatically, it will reflect the change.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -737,7 +737,7 @@ jobs:
 
   build_win:
     name: Windows
-    runs-on: windows-2019
+    runs-on: windows-2022
```
Comment (ReenigneArcher) [inline .github/workflows/CI.yml:740]: Is this change necessary for your PR? I'd prefer if this change were in a separate PR.

File: third-party/nvapi-windows-amd64/include/NvApiDriverSettings.c
Comment (ReenigneArcher) [inline third-party/nvapi-windows-amd64/include/NvApiDriverSettings.c:1]: What is the original source of these third-party files?

File: tools/nvprefs/nvprefs.cpp (hunk 2 above)
Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp]: Sorry for seemingly ignoring you, just had to basically rewrite a lot of stuff and this question was hardly a priority at that moment.

After giving it more thought, I think the most logical way is to follow what NVIDIA control panel itself does. Whether after `NvAPI_DRS_SetCurrentGlobalProfile()` it switches to that profile, or sticks to the base one. Will test this and possibly make modifications accordingly.

Comment (Lucide) [reply tools/nvprefs/nvprefs.cpp]: No worries, mine were just suggestions. If the code is sound, corrections will be trivial should issues arise.

Comment (ns6089) [reply tools/nvprefs/nvprefs.cpp]: Just tested, NVIDIA control panel only displays and modifies base profile even after custom one is set, and that custom profile properly inherits from the base one. So unless custom global profile explicitly overrides the settings we change we should be fine, and I'd rather not change settings that are essentially hidden from the user. Sticking to the base profile for now, unless some particular use-case arises where it is not enough.

**Comment by ns6089** [comment]
As far as I can tell, this should be ready to merge.

**Comment by ReenigneArcher** [comment]
> As far as I can tell, this should be ready to merge.

I think we just need to deal with the third party files. I'm reviewing on my phone so I could be missing something.

https://github.com/LizardByte/Sunshine/pull/1308#discussion_r1240469153

**Comment by ns6089** [comment]
> I think we just need to deal with the third party files. I'm reviewing on my phone so I could be missing something.
> 
> [#1308 (comment)](https://github.com/LizardByte/Sunshine/pull/1308#discussion_r1240469153)

That comment is now outdated, I found the opensource version NVIDIA distributes. To quote:
```
This release contains a version of nvapi.h, provided under MIT license, to enable open-source
re-implementations of NVAPI for Windows emulation environments.
```
So no need for binary libraries, standalone tools, and everything is getting called from sunshine.exe

`third-party/nvapi-opensource/readme.txt` contains the url for possible updates and everything.
But these updates won't be really needed unless we will need new functions, everything else is backwards-compatible.
It's only headers basically.

**Comment by ReenigneArcher** [comment]
I'd prefer to add them as submodules if it's possible.

**Comment by ns6089** [comment]
> I'd prefer to add them as submodules if it's possible.

Feel free place it anywhere you want, I will modify the PR. But I don't have the powuh to create repo/branch/whatever myself.

**Comment by Lucide** [comment]
If the sunshine service is launched at startup and left idling in background, and the high performance setting is set on `main()`, does that mean the gpu will be in a constant high performance/power consumption state?

**Comment by ns6089** [comment]
> If the sunshine service is launched at startup and left idling in background, and the high performance setting is set on `main()`, does that mean the gpu will be in a constant high performance/power consumption state?
> 
> Does the setting apply only on subsequent launches of sunshine?

A little bit of context. These settings are stored in `c:\ProgramData\NVIDIA Corporation\Drs\nvdrsdb0.bin`. After an application creates 3d context, it loads nvidia libraries and these libraries read that file in application memory space.

Since sunshine doesn't use any 3d contexts when not actively streaming, power states will not be constantly applied, and the setting will be applied even on the first launch (when streaming starts).

**Comment by Lucide** [comment]
Oh, I see!
Does it also respect context "destruction"? i.e., does the driver unload the app settings when the stream ends?

**Comment by ns6089** [comment]
> Oh, I see! Does it also respect context "destruction"? i.e., does the driver unload the app settings when the stream ends?

If it's power states no longer being applied, then yes. If it's settings being re-read on the next 3d context creation, I haven't tested it, but I don't see why they shouldn't. But since it tied to library load/unload, you might need to destroy every last 3d context for reload to happen.

**Comment by Lucide** [comment]
It's the above, I was worrying about making the changes as stream-local as possible. So for the app profile this happens implicitly.

Does it make sense to move the base profile changes explicitly at stream start/end too?

promise I won't bother you no further😶‍🌫️

**Comment by ns6089** [comment]
> Does it make sense to move the base profile changes explicitly at stream start/end too?

No worries, it's a valid question. We change the *base* profile because the api we use for capturing doesn't work well on opengl/vulkan games, and we need to change how their presentation works to sidestep it. And the game can be launched before streaming request comes.

File: third-party/nvapi-opensource/readme.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+Opensource version of nvapi
+Updates can be found at https://download.nvidia.com/XFree86/nvapi-open-source-sdk/
```
Comment (ReenigneArcher) [inline third-party/nvapi-opensource/readme.txt:2]: Okay, I see it doesn't really seem to be available in any updated git repo (that I could find anyway). Could we just download and extract this with CMakeLists? I'm not really a fan of including other source code directly if it can be avoided.

https://cmake.org/cmake/help/latest/command/file.html#transfer
https://cmake.org/cmake/help/latest/command/file.html#archive-extract

**Review by ReenigneArcher — CHANGES_REQUESTED** [review]

File: third-party/nvapi-opensource/readme.txt (hunk 4 above)
Comment (ns6089) [reply third-party/nvapi-opensource/readme.txt:2]: But let's look upon this objectively

What cons is this going to bring?
- We'll be effectively breaking repo history once they change the url, and it's going to happen sooner or later
- It can cause immense annoyance in places with spotty internet (hey, I like taking my laptop to the mountains)

What pros?
- I don't see any

Comment (ns6089) [reply third-party/nvapi-opensource/readme.txt:2]: It's just, I think we're spending time trying to solve a problem that wasn't a problem to begin with. And if I start looking into that `archive-extract`, it will introduce actual problems that will require actual solving and more back and forth communication.

Comment (Lucide) [reply third-party/nvapi-opensource/readme.txt:2]: If this acts as the regular nvapi drs then it's not even a real library at all, it's just a loader for the real library inside the driver.
It's made to be forward compatible forever, it's not something to keep updated.
The Nvidia drs dev guide is from 2011 and points this out.

File: src/platform/windows/nvprefs/undo_file.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  std::optional<undo_file_t>
+  undo_file_t::create_new_file(std::filesystem::path file_path) {
+    undo_file_t file;
+    file.file_handle.reset(CreateFileW(file_path.c_str(), GENERIC_WRITE | STANDARD_RIGHTS_ALL, 0, nullptr, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL));
+
+    if (file.file_handle) {
+      // give GENERIC_READ, GENERIC_WRITE and DELETE permissions to Users group
+      if (relax_permissions(file.file_handle.get()) != 0) {
```
Comment (cgutman) [inline src/platform/windows/nvprefs/undo_file.cpp:66]: What's the thinking here for allowing the Users group to write this file (or really for any user/process besides us to write it)?

File: src/platform/windows/nvprefs/nvprefs_interface.cpp [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    }
+    unload();
+  }
+
+  bool
+  nvprefs_interface::load() {
+    if (!pimpl->loaded) {
+      // Check %ProgramData% variable, need it for storing undo file
+      wchar_t program_data_env[MAX_PATH];
+      auto get_env_result = GetEnvironmentVariableW(L"ProgramData", program_data_env, MAX_PATH);
```
Comment (cgutman) [inline src/platform/windows/nvprefs/nvprefs_interface.cpp:43]: Maybe we should use `%LOCALAPPDATA%` instead of `%ProgramData%` to keep the undo state for installer and portable versions of Sunshine apart.

Comment (ns6089) [reply src/platform/windows/nvprefs/nvprefs_interface.cpp:43]: The NVIDIA settings themselves are global and stored in %ProgramData% file. If we split out undo states, one can pollute the global state with no way for the other one to know about it.

File: src/platform/windows/nvprefs/undo_file.cpp (hunk 5 above)
Comment (ns6089) [reply src/platform/windows/nvprefs/undo_file.cpp:66]: It's really GENERIC_READ and DELETE that's is required logic-wise - read file, apply undo, delete file. But GENERIC_WRITE is needed for locking the file during the process.

"Us" is defined here by file path, not user or process. Since we have portable version that can be be run in parallel and as any user, and then can crash and leave the undo file with settings not reverted. If NVIDIA state is global and can be changed by anyone, then the undo file must be global too, I don't think there's a way around it.

File: third-party/nvapi-opensource/readme.txt (hunk 4 above)
Comment (ns6089) [reply third-party/nvapi-opensource/readme.txt:2]: That's the thing, the files in question are not code, but constants. And we don't really need to update them unless we would want to change some newly added parameters. Which we won't be needing for years, most likely ever. And if we can't find the parameter in question, that's the sure sign to update. Which takes 3 minutes since it's trivial and kept in one place. You will spend order of magnitudes more time on automation.

**Review by cgutman — APPROVED** [review]

**Comment by cgutman** [comment]
It looks like we need to ignore the third-party files for clang-format. Can you try adding a `.clang-format-ignore` with `third-party/*` and see if that fixes it?

**Comment by ReenigneArcher** [comment]
> 

Due to how the workflow finds the files, that probably won't work (assumption).

https://github.com/LizardByte/Sunshine/blob/50f353d183e2a6e22fd69f4320691c27d87c8b80/.github/workflows/cpp-lint.yml#L31

I like the idea though so will work on a way to implement that.

**Comment by ReenigneArcher** [comment]
@cgutman This should work. https://github.com/LizardByte/Sunshine/pull/1449
